runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp

```diff
@@ -693,26 +693,40 @@ Java_java_lang_invoke_MethodHandleNatives_expand(JNIEnv *env, jclass clazz, jobj
 }
 
 /**
- * static native MemberName resolve(MemberName self, Class<?> caller,
- *      boolean speculativeResolve) throws LinkageError, ClassNotFoundException;
+ * [JDK8] static native MemberName resolve(MemberName self, Class<?> caller)
+ * 	             throws LinkageError, ClassNotFoundException;
  *
- * Resolve the method/field represented by the MemberName's symbolic data (name & type & defc) with the supplied caller
- * Store the resolved Method/Field's JNI-id in vmindex, field offset / method pointer in vmtarget
+ * [JDK11] static native MemberName resolve(MemberName self, Class<?> caller,
+ * 	             boolean speculativeResolve) throws LinkageError, ClassNotFoundException;
+ *
+ * [JDK16+] static native MemberName resolve(MemberName self, Class<?> caller, int lookupMode,
+ * 	             boolean speculativeResolve) throws LinkageError, ClassNotFoundException;
+ *
+ * Resolve the method/field represented by the MemberName's symbolic data (name & type & defc)
+ * with the supplied caller. Store the resolved Method/Field's JNI-id in vmindex, field offset
+ * or method pointer in vmtarget.
  *
  * If the speculativeResolve flag is not set, failed resolution will throw the corresponding exception.
- * If the resolution failed with no exception:
- * Throw NoSuchFieldError for field MemberName
- * Throw NoSuchMethodError for method/constructor MemberName
- * Throw LinkageError for other
+ *
+ * If the resolution failed with no exception,
+ *   - Throw NoSuchFieldError for field MemberName issues.
+ *   - Throw NoSuchMethodError for method/constructor MemberName issues.
+ *   - Throw LinkageError for other issues.
  */
-#if JAVA_SPEC_VERSION >= 16
-jobject JNICALL
-Java_java_lang_invoke_MethodHandleNatives_resolve(JNIEnv *env, jclass clazz, jobject self, jclass caller, jint lookupMode, jboolean speculativeResolve)
-#else /* JAVA_SPEC_VERSION >= 16 */
 jobject JNICALL
-Java_java_lang_invoke_MethodHandleNatives_resolve(JNIEnv *env, jclass clazz, jobject self, jclass caller, jboolean speculativeResolve)
-#endif /* JAVA_SPEC_VERSION >= 16 */
-{
+Java_java_lang_invoke_MethodHandleNatives_resolve(
+#if JAVA_SPEC_VERSION == 8
+                                                  JNIEnv *env, jclass clazz, jobject self, jclass caller
+#elif JAVA_SPEC_VERSION == 11 /* JAVA_SPEC_VERSION == 8 */
+                                                  JNIEnv *env, jclass clazz, jobject self, jclass caller,
+                                                  jboolean speculativeResolve
+#elif JAVA_SPEC_VERSION >= 16 /* JAVA_SPEC_VERSION == 11 */
+                                                  JNIEnv *env, jclass clazz, jobject self, jclass caller,
+                                                  jint lookupMode, jboolean speculativeResolve
+#else /* JAVA_SPEC_VERSION == 8 */
+#error Unsupported Java_java_lang_invoke_MethodHandleNatives_resolve
```

Comment:
same here, in what scenario would this fail?

Reply:
Will fail for gaps in `ifdef`: JDKs 9, 10, 12, 13, 14, 15 and 16. Although unsupported, the `#else` should be handled; similar to introducing a `default` for `switch` statements.

Reply:
I see. If you attempt to build those version you will have other problems, so I dont think its necessary to add the error case

Reply:
Removed.